src/+mpecopt/Solver.m

```diff
@@ -1157,5 +1334,454 @@ function create_lpec_functions(obj)
+                nabla_f_k = full(mpec_casadi.nabla_f_fun(x_k,p0));
+                active_set_estimate_k = find_active_sets_piece_nlp(x_k,nabla_f_k,y_lpec_k_l,dims,opts,opts.tol_active);
+                ubx_bnlp_k(dims.ind_x1(active_set_estimate_k.I_0_plus)) = 0;
+                ubx_bnlp_k(dims.ind_x2(active_set_estimate_k.I_plus_0)) = 0;
+                ubx_bnlp_k(dims.ind_x1(active_set_estimate_k.I_00)) = 0;
+                ubx_bnlp_k(dims.ind_x2(active_set_estimate_k.I_00)) = 0;
+                opts.piece_nlp_strategy = initial_strategy;
+                % end
+                t_nlp_start = tic;
+                results_nlp = mpec_casadi.solver('x0',x_k,'p', p0, 'lbx', lbx_bnlp_k, 'ubx', ubx_bnlp_k,'lbg', lbg_tnlp_k, 'ubg', ubg_tnlp_k);
+                cpu_time_nlp_k_l = toc(t_nlp_start);
+                x_k_multi = full(results_nlp.x);
+                lambda_x_k = full(results_nlp.lam_x);
+                [stats.multiplier_based_stationarity, ~] = determine_multipliers_based_stationary_point(x_k_multi,lambda_x_k,dims,opts);
+            end
+
+            % Debug falure of stationary point computation
+            if ~strcmp(multiplier_based_stationarity_debug,'X') && opts.debug_mode_on &&  ~isequal(multiplier_based_stationarity_debug,stats.multiplier_based_stationarity)
+                keyboard;
+            end
+
+            % eval stats at solution
+            % x_k = x_k(1:dims.n_primal_non_lifted);
+            solution.x = x_k(dims.map_w(1:dims.n_primal_non_lifted));
+            solution.x_lifted = x_k(dims.map_w);
+            solution.f = full(mpec_casadi.f_fun(x_k,p0));
+            solution.x1_opt = x_k(dims.ind_x1);
+            solution.x2_opt = x_k(dims.ind_x2);
+            stats.h_std = full(mpec_casadi.h_std_fun(x_k,p0));
+            stats.comp_res = full(mpec_casadi.h_comp_fun(x_k,p0));
+            stats.h_total = full(mpec_casadi.h_total_fun(x_k,p0));
+            stats.total_outer_iterations = k;
+
+            stats.n_biactive = sum(x_k(dims.ind_x1)+x_k(dims.ind_x2) < 2*opts.tol_active );
+            try
+                stats.lambda_g_opt = full(results_nlp.lam_g(dims.map_g));
+                stats.lambda_x_opt = full(results_nlp.lam_x(dims.map_x));
+                stats.n_active_ineq = sum(abs(lambda_g_opt(ind_g_ineq))>opts.tol);
+                stats.n_active_box = sum(abs((lambda_x_opt))>opts.tol & lbx_bnlp_k~=ubx_bnlp_k);
+                stats.n_box = sum(lbx_bnlp_k~=ubx_bnlp_k);
+                stats.n_box_simple = sum(lbx_bnlp_k==ubx_bnlp_k);
+            catch
+                stats.n_active_ineq = nan;
+                stats.n_active_box = nan;
+                stats.n_box = sum(solver_initialization.lbx~=solver_initialization.ubx);
+                stats.n_box_simple = sum(solver_initialization.lbx==solver_initialization.ubx);
+            end
+        end
+
+    end
+end
+
+
+function X = all_combinations(varargin)
```

Comment:
please add some comments what are these specalized functons doing and what are they needed for?